Master Lightning encountered an error when importing data to an older version of TiDB: "failed to process the router requests".

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

1. lightning version: 
```
bash-5.1# /tidb-lightning -V
Release Version: v9.0.0-beta.2.pre-1585-gb6200ce8d0
Git Commit Hash: b6200ce8d00986696ee0bba25fb4c0b9c792d993
Git Branch: HEAD
Go Version: go1.25.8
UTC Build Time: 2026-04-20 04:41:06
Race Enabled: false
Kernel Type: Classic
```
2. use lightning import data to v8.5.6 TiDB
```
/tidb-lightning -tidb-port 4000 -tidb-user root -backend "local" -sorted-kv-dir "/tmp/sorted-kv-dir" -d "s3://nfs/jianxu/test-data?access-key=xxx&secret-access-key=xxx&endpoint=http://minio.pingcap.net:9001&force-path-style=true" 
-pd-urls "upstream-pd.xxxxx:2379" -tidb-host "upstream-tidb.xxx"
```

### 2. What did you expect to see? (Required)
import success

### 3. What did you see instead (Required)
```
[2026/04/20 09:43:32.188 +00:00] [ERROR] [client.go:796] ["[router] failed to process the router requests"] [stream-url=http://upstream-pd-0.upstream-pd-peer.brie-version-comp-tps-8136540-1-398.svc.cluster.local:2379] [error="rpc error: code = Unimplemented desc = unknown method QueryRegion for service pdpb.PD"] [stack="github.com/tikv/pd/client/clients/router.(*Cli).handleProcessRequestError\n\t/root/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20260404141330-8a6813497b52/clients/router/client.go:796\ngithub.com/tikv/pd/client/clients/router.(*Cli).dispatcher\n\t/root/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20260404141330-8a6813497b52/clients/router/client.go:626"]
[2026/04/20 09:43:32.188 +00:00] [WARN] [client.go:517] ["wait for split failed"] [regionID=150] [error="rpc error: code = Unimplemented desc = unknown method QueryRegion for service pdpb.PD"] [errorVerbose="rpc error: code = Unimplemented desc = unknown method QueryRegion for service pdpb.PD\ngithub.com/tikv/pd/client/clients/router.(*Request).wait\n\t/root/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20260404141330-8a6813497b52/clients/router/request.go:70\ngithub.com/tikv/pd/client/clients/router.(*Cli).GetRegionByID\n\t/root/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20260404141330-8a6813497b52/clients/router/request.go:114\ngithub.com/tikv/pd/client.(*client).GetRegionByID\n\t/root/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20260404141330-8a6813497b52/client.go:825\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).GetRegionByID\n\t/workspace/source/tidb/br/pkg/restore/split/client.go:290\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).hasHealthyRegion\n\t/workspace/source/tidb/br/pkg/restore/split/client.go:547\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).waitRegionsSplit.func1\n\t/workspace/source/tidb/br/pkg/restore/split/client.go:514\ngithub.com/pingcap/tidb/br/pkg/utils.WithRetryReturnLastErr\n\t/workspace/source/tidb/br/pkg/utils/retry.go:86\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).waitRegionsSplit\n\t/workspace/source/tidb/br/pkg/restore/split/client.go:508\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).SplitWaitAndScatter\n\t/workspace/source/tidb/br/pkg/restore/split/client.go:684\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).SplitKeysAndScatter.func1.1\n\t/workspace/source/tidb/br/pkg/restore/split/client.go:619\ngithub.com/pingcap/tidb/pkg/util.(*WorkerPool).ApplyOnErrorGroup.func1\n\t/workspace/source/tidb/pkg/util/worker_pool.go:81\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/root/go/pkg/mod/golang.org/x/sync@v0.20.0/errgroup/errgroup.go:93\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1693"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
mysql> SELECT TIDB_VERSION()\G
*************************** 1. row ***************************
TIDB_VERSION(): Release Version: v8.5.6
Edition: Community
Git Commit Hash: ae18096e023780bb56bfce33698abec0d4640d0a
Git Branch: HEAD
UTC Build Time: 2026-04-14 07:15:36
GoVersion: go1.25.8
Race Enabled: false
Check Table Before Drop: false
Store: tikv
1 row in set (0.04 sec)

--- a/br/pkg/task/BUILD.bazel
+++ b/br/pkg/task/BUILD.bazel
@@ -126,7 +126,7 @@ go_test(
     ],
     embed = [":task"],
     flaky = True,
-    shard_count = 47,
+    shard_count = 49,
     deps = [
         "//br/pkg/backup",
         "//br/pkg/config",

--- a/br/pkg/task/operator/BUILD.bazel
+++ b/br/pkg/task/operator/BUILD.bazel
@@ -79,7 +79,7 @@ go_test(
         "//br/pkg/task",
         "//pkg/objstore",
         "//pkg/objstore/storeapi",
-        "@com_github_stretchr_testify//require",
         "@com_github_spf13_pflag//:pflag",
+        "@com_github_stretchr_testify//require",
     ],
 )

--- a/lightning/pkg/importer/BUILD.bazel
+++ b/lightning/pkg/importer/BUILD.bazel
@@ -96,6 +96,7 @@ go_library(
         "@com_github_tikv_client_go_v2//util",
         "@com_github_tikv_pd_client//:client",
         "@com_github_tikv_pd_client//http",
+        "@com_github_tikv_pd_client//opt",
         "@com_github_tikv_pd_client//pkg/caller",
         "@com_github_tikv_pd_client//pkg/retry",
         "@io_etcd_go_etcd_client_v3//:client",

--- a/lightning/pkg/importer/import.go
+++ b/lightning/pkg/importer/import.go
@@ -384,6 +384,7 @@ func NewImportControllerWithPauser(
 		}
 
 		addrs := strings.Split(cfg.TiDB.PdAddr, ",")
+		// Disable router QueryRegion streams so newer Lightning can import into older PD clusters.
 		pdClientOptions := append(ingestctrl.PDClientOptions(), opt.WithEnableRouterClient(false))
 		pdCli, err = pd.NewClientWithAPIContext(ctx, apiContext, componentName, addrs, tls.ToPDSecurityOption(), pdClientOptions...)
 		if err != nil {
@@ -1212,7 +1213,10 @@ const (
 func (rc *Controller) keepPauseGCForDupeRes(ctx context.Context) (<-chan struct{}, error) {
 	tlsOpt := rc.tls.ToPDSecurityOption()
 	addrs := strings.Split(rc.cfg.TiDB.PdAddr, ",")
-	pdCli, err := pd.NewClientWithAPIContext(ctx, rc.apiContext, componentName, addrs, tlsOpt)
+	// Disable router QueryRegion streams for older PD clusters during duplicate resolution.
+	pdCli, err := pd.NewClientWithAPIContext(
+		ctx, rc.apiContext, componentName, addrs, tlsOpt, opt.WithEnableRouterClient(false),
+	)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}

--- a/lightning/pkg/importer/import.go
+++ b/lightning/pkg/importer/import.go
@@ -71,6 +71,7 @@ import (
 	kvutil "github.com/tikv/client-go/v2/util"
 	pd "github.com/tikv/pd/client"
 	pdhttp "github.com/tikv/pd/client/http"
+	"github.com/tikv/pd/client/opt"
 	"github.com/tikv/pd/client/pkg/caller"
 	"github.com/tikv/pd/client/pkg/retry"
 	clientv3 "go.etcd.io/etcd/client/v3"
@@ -383,7 +384,8 @@ func NewImportControllerWithPauser(
 		}
 
 		addrs := strings.Split(cfg.TiDB.PdAddr, ",")
-		pdCli, err = pd.NewClientWithAPIContext(ctx, apiContext, componentName, addrs, tls.ToPDSecurityOption(), ingestctrl.PDClientOptions()...)
+		pdClientOptions := append(ingestctrl.PDClientOptions(), opt.WithEnableRouterClient(false))
+		pdCli, err = pd.NewClientWithAPIContext(ctx, apiContext, componentName, addrs, tls.ToPDSecurityOption(), pdClientOptions...)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
@@ -448,6 +450,8 @@ func NewImportControllerWithPauser(
 			raftKV2SwitchModeDuration = cfg.Cron.SwitchMode.Duration
 		}
 		backendConfig := ingestctrl.NewBackendConfig(cfg, maxOpenFiles, p.KeyspaceName, p.ResourceGroupName, p.TaskType, raftKV2SwitchModeDuration)
+		// Use legacy PD region RPCs so newer Lightning can import into older clusters.
+		backendConfig.DisablePDClientRouterClient = true
 		backendObj, err = ingestctrl.NewBackend(ctx, tls, backendConfig, pdCliForTiKV)
 		if err != nil {
 			return nil, common.NormalizeOrWrapErr(common.ErrUnknown, err)

--- a/pkg/ingestor/ingestctrl/local.go
+++ b/pkg/ingestor/ingestctrl/local.go
@@ -487,7 +487,7 @@ type BackendConfig struct {
 	ResourceGroupName         string
 	TaskType                  string
 	RaftKV2SwitchModeDuration time.Duration
-	// DisablePDClientRouterClient keeps region lookups on legacy PD RPCs.
+	// DisablePDClientRouterClient uses legacy PD region RPCs for callers that must work with older PD clusters.
 	DisablePDClientRouterClient bool
 	// whether disable automatic compactions of pebble db of engine.
 	// deduplicate pebble db is not affected by this option.

--- a/pkg/ingestor/ingestctrl/local.go
+++ b/pkg/ingestor/ingestctrl/local.go
@@ -487,6 +487,8 @@ type BackendConfig struct {
 	ResourceGroupName         string
 	TaskType                  string
 	RaftKV2SwitchModeDuration time.Duration
+	// DisablePDClientRouterClient keeps region lookups on legacy PD RPCs.
+	DisablePDClientRouterClient bool
 	// whether disable automatic compactions of pebble db of engine.
 	// deduplicate pebble db is not affected by this option.
 	// see DisableAutomaticCompactions of pebble.Options for more details.
@@ -845,7 +847,11 @@ func (local *Backend) getTiKVClient(ctx context.Context) (*tikvclient.KVStore, e
 	// the lifecycle of input PD client, while the PD client inside this is
 	// managed outside.
 	apiContext := keyspace.BuildAPIContext(local.KeyspaceName)
-	pdCliForTiKV, err := newPDClient(ctx, apiContext, caller.Component("lightning-local-backend"), local.pdAddrs, pdSecurityOption(local.tls), PDClientOptions()...)
+	pdClientOptions := PDClientOptions()
+	if local.DisablePDClientRouterClient {
+		pdClientOptions = append(pdClientOptions, opt.WithEnableRouterClient(false))
+	}
+	pdCliForTiKV, err := newPDClient(ctx, apiContext, caller.Component("lightning-local-backend"), local.pdAddrs, pdSecurityOption(local.tls), pdClientOptions...)
 	if err != nil {
 		_ = spkv.Close()
 		return nil, common.ErrCreateKVClient.Wrap(err).GenWithStackByArgs()

--- a/pkg/ingestor/ingestctrl/local_test.go
+++ b/pkg/ingestor/ingestctrl/local_test.go
@@ -2475,6 +2475,7 @@ func TestGetDupeControllerInitializesTiKVClientLazily(t *testing.T) {
 		pdClientCalls++
 		require.Equal(t, pd.NewAPIContextV1(), apiContext)
 		option := opt.NewOption()
+		option.SetEnableRouterClient(true)
 		for _, clientOpt := range opts {
 			clientOpt(option)
 		}
@@ -2513,6 +2514,28 @@ func TestGetDupeControllerInitializesTiKVClientLazily(t *testing.T) {
 	require.Equal(t, 1, rpcClientCalls)
 	require.Equal(t, 1, kvStoreCalls)
 	require.False(t, routerClientEnabled)
+	require.False(t, inputPDCli.closed)
+	require.True(t, tikvPDCli.closed)
+	require.Nil(t, b.tikvCli)
+
+	pdClientCalls, safePointKVCalls, rpcClientCalls, kvStoreCalls = 0, 0, 0, 0
+	routerClientEnabled = false
+	inputPDCli = &mockPdClient{}
+	tikvPDCli = &mockPdClient{}
+	b = &Backend{
+		pdCli:     inputPDCli,
+		pdAddrs:   []string{"127.0.0.1:2379"},
+		tls:       &common.TLS{},
+		tikvCodec: keyspace.CodecV1,
+	}
+	dupeController, err = b.GetDupeController(context.Background(), 1, nil)
+	require.Nil(t, dupeController)
+	require.ErrorContains(t, err, "mock kv store error")
+	require.Equal(t, 1, pdClientCalls)
+	require.Equal(t, 1, safePointKVCalls)
+	require.Equal(t, 1, rpcClientCalls)
+	require.Equal(t, 1, kvStoreCalls)
+	require.True(t, routerClientEnabled)
 	require.False(t, inputPDCli.closed)
 	require.True(t, tikvPDCli.closed)
 	require.Nil(t, b.tikvCli)

--- a/pkg/ingestor/ingestctrl/local_test.go
+++ b/pkg/ingestor/ingestctrl/local_test.go
@@ -2468,11 +2468,17 @@ func TestGetDupeControllerInitializesTiKVClientLazily(t *testing.T) {
 	}()
 
 	var pdClientCalls, safePointKVCalls, rpcClientCalls, kvStoreCalls int
+	var routerClientEnabled bool
 	inputPDCli := &mockPdClient{}
 	tikvPDCli := &mockPdClient{}
-	newPDClient = func(_ context.Context, apiContext pd.APIContext, _ caller.Component, _ []string, _ pd.SecurityOption, _ ...opt.ClientOption) (pd.Client, error) {
+	newPDClient = func(_ context.Context, apiContext pd.APIContext, _ caller.Component, _ []string, _ pd.SecurityOption, opts ...opt.ClientOption) (pd.Client, error) {
 		pdClientCalls++
 		require.Equal(t, pd.NewAPIContextV1(), apiContext)
+		option := opt.NewOption()
+		for _, clientOpt := range opts {
+			clientOpt(option)
+		}
+		routerClientEnabled = option.GetEnableRouterClient()
 		return tikvPDCli, nil
 	}
 	newEtcdSafePointKV = func(_ []string, _ *tls.Config, _ ...tikv.SafePointKVOpt) (tikv.SafePointKV, error) {
@@ -2494,6 +2500,9 @@ func TestGetDupeControllerInitializesTiKVClientLazily(t *testing.T) {
 		pdAddrs:   []string{"127.0.0.1:2379"},
 		tls:       &common.TLS{},
 		tikvCodec: keyspace.CodecV1,
+		BackendConfig: BackendConfig{
+			DisablePDClientRouterClient: true,
+		},
 	}
 
 	dupeController, err := b.GetDupeController(context.Background(), 1, nil)
@@ -2503,6 +2512,7 @@ func TestGetDupeControllerInitializesTiKVClientLazily(t *testing.T) {
 	require.Equal(t, 1, safePointKVCalls)
 	require.Equal(t, 1, rpcClientCalls)
 	require.Equal(t, 1, kvStoreCalls)
+	require.False(t, routerClientEnabled)
 	require.False(t, inputPDCli.closed)
 	require.True(t, tikvPDCli.closed)
 	require.Nil(t, b.tikvCli)

